### PR TITLE
executor: make point updates go short path

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

make point update cached like point select decoupled with `prepared plan cache`

- sysbench `oltp_update_index` and `oltp_update_non_index`
- tpcc `benchmarksql`
will get benefit from this

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

Release note

### CODE REVIEW

**Comment by zz-jason** [comment]
@cfzjywxk please add some proper labels for this PR.

File: session/session.go [hunk 1]
```diff
@@ -1214,12 +1224,23 @@ func (s *session) IsCachedExecOk(ctx context.Context, prepared *ast.Prepared) (b
...
+		ok = true
+	default:
+		ok = false
+	}
+	return ok, err
 }
 
 // ExecutePreparedStmt executes a prepared statement.
 func (s *session) ExecutePreparedStmt(ctx context.Context, stmtID uint32, args []types.Datum) (sqlexec.RecordSet, error) {
+	s.PrepareTxnCtx(ctx)
```
Comment (jackysp) [inline session/session.go]: why change this?

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -308,6 +305,33 @@ func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context,
 	return err
 }
 
+// tryCachePointPlan will try to cache point execution plan, there may be some
+// short paths for these executions, currently "point select" and "point update"
+func (e *Execute) tryCachePointPlan(ctx context.Context, sctx sessionctx.Context,
+	prepared *ast.Prepared, is infoschema.InfoSchema, p Plan) error {
+	var ok bool
```
Comment (jackysp) [inline planner/core/common_plans.go]:
```suggestion
	var (
	    ok bool
	    err error
	)
```

File: planner/core/common_plans.go [hunk 3]
```diff
@@ -308,6 +305,33 @@ func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context,
...
+	var err error
+	switch p.(type) {
+	case *PointGetPlan:
+		ok, err = IsPointGetWithPKOrUniqueKeyByAutoCommit(sctx, p)
+	case *Update:
+		ok, err = IsPointUpdateByAutoCommit(sctx, p)
```
Comment (jackysp) [inline planner/core/common_plans.go]: I think it is better to check the err here.

File: planner/core/common_plans.go [hunk 4]
```diff
@@ -895,3 +919,30 @@ func IsPointGetWithPKOrUniqueKeyByAutoCommit(ctx sessionctx.Context, p Plan) (bo
 		return false, nil
 	}
 }
+
+// IsPointUpdateByAutoCommit checks if plan p is point update and is in autocommit context
+func IsPointUpdateByAutoCommit(ctx sessionctx.Context, p Plan) (bool, error) {
```
Comment (jackysp) [inline planner/core/common_plans.go]: There are some duplicated code with IsPointGetWithPKOrUniqueKeyByAutoCommit.

File: session/session.go (hunk 1 above)
Comment (cfzjywxk) [reply session/session.go]: both cached execution and compile execution path need this, put here to make it simple

File: session/session.go [hunk 5]
```diff
@@ -1214,12 +1224,23 @@ func (s *session) IsCachedExecOk(ctx context.Context, prepared *ast.Prepared) (b
 		return false, nil
 	}
 	// maybe we'd better check cached plan type here, current
-	// only point select will be cached, see "getPhysicalPlan" func
-	return true, nil
+	// only point select/update will be cached, see "getPhysicalPlan" func
+	var ok bool
+	var err error
+	switch prepared.CachedPlan.(type) {
+	case *plannercore.PointGetPlan:
```
Comment (imtbkcat) [inline session/session.go]:
```suggestion
	case *plannercore.PointGetPlan, *plannercore.Update:
	    ok = true
```

**Comment by cfzjywxk** [comment]
@jackysp  @imtbkcat PTAL

**Comment by jackysp** [comment]
PTAL @lysu

**Pull request merged**